docs/user-guide/universal-configuration.rst

```diff
@@ -0,0 +1,410 @@
+and is to be avoided.
+
+
+.. _universal_configuration_advanced:
+
+Advanced list behavior
+----------------------
+
+In this section, we will go over some of the more complex aspects of Universal
+Configuration.
+
+Most of these are not as universally required when using the majority of roles
+and when they are the role documentation will give you fair warning.
+
+
+.. _universal_configuration_option:
+
+Controlled merging
+~~~~~~~~~~~~~~~~~~
+
+Items use ``name`` as a unique key. The underlying filter *does* allow a role
+to change the name of the field used for this purpose, by providing a ``name``
+argument,  but for consistency it is generally discouraged unless there's a
+*really* compelling reason to do so.
+
+This has the implication that configuration options which *may* appear multiple
+times in valid configuration would override each other if naively implemented.
+
+The ``option`` field exists for this purpose:
+
+.. code-block:: yaml
+
+   - name: 'timeout'
+     value: 2 * 60 * 60
+
+   - name: 'my first include'
+     option: 'include'
+     value: '/etc/fizz'
+
+   - name: 'my other include'
+     option: 'include'
+     value: '/etc/buzz'
+
+In this example, our two ``include`` statements can coexist and be modified
+as expected by later items targeting their ``name`` field.
+
+The role can then loop through the resulting list in its templates with a
+single statement like the one below:
+
+.. code-block:: jinja
+
+  {{ '{} = {}'.format((item.option | d(item.name)), item.value) }}
+
+
+.. _universal_configuration_advanced_weight:
+
+Weighing and anchoring
+~~~~~~~~~~~~~~~~~~~~~~
+
+Under the hood, the configuration filters populate an ``id`` field for each
+item in multiples of 10, starting from 0.
+
+An item's ``weight`` is added to that ``id`` to come up with the final sorting
+order, stored in a field called ``.real_weight``.
+
+.. note::
+
+   Take note that the initial order of the list items matters as much as
+   the ``weight`` you provide.
+
+   The specifics of the weight behavior can be counterintuitive and are
+   currently under review. Don't build too intricate orderings that you
+   cannot afford to rewrite, and watch this space!
+
+
+In more complex scenarios, the ``copy_id_from`` key allows us to reference
+another list item by ``name``. Its ``real_weight`` will then be calculated
+based on that referenced item's ``id``.
+
+
+.. _universal_configuration_recursion:
+
+Recursion
+~~~~~~~~~
+
+There are two different recursion behaviors for item fields, differentiated by
+whether values of multiple fields will be merged or not.
+
+If an item's ``value`` field contains a list, the configuration filters will
+recurse into it, so any of the documented configuration syntax can be used in
+it as well.
+
+When merging items with the same ``name`` whose ``value`` fields contain lists,
+those lists **will not be merged**.
+Only the last appearing ``value`` will be used.
```

Comment:
That's true for the `parse_kv_items` filter, but not for the `parse_kv_config` filter. In the latter case, values of lists are additive so that multiple elements can be added afterwards. If you want to "reset" a value, you need to use a string instead. This behaviour is used for example in the [`postfix` role](https://docs.debops.org/en/master/ansible/roles/postfix/defaults-maincf.html#lists-are-merged-together) to manage `/etc/postfix/main.cf` lists.

Reply:
Uh this makes my brain hurt. The end user cannot differentiate between the two filters without peeking at the source, can they? I was hoping to reveal the fact that we're talking about two filters only in the role dev doc. :(

Reply:
I guess you can still do that later. I think that you can describe the behaviour of `parse_kv_config` in the user guide (list values are merged) because it will be commonly encountered in the correct places, for example inside configuration files. And you can leave the specifics of `parse_kv_items` o the dev doc.

Reply:
I'll note the difference between `parse_kv_items` and `_config` by mentioning both can happen and to check the role doc, for now, and I'll think about unifying behavior in the `config` rewrite.

Can I consider `_config` to be the default unless otherwise noted?

Reply:
Sure, `parse_kv_config` can be used as the default behaviour.

The `_config` is for single configuration files`, `_items` is used outside in the task to handle multiple configuration files in `conf.d/` style setups where the contents are handled by `_config`. That's the common use case. Different behaviour of these filters allows for more complex setups, for example `/etc/postfix/master.cf` where `parse_kv_items` is used in a single configuration file which has muti-level configuration. But that's an exception.